prelude/slang-cuda-prelude.h

```diff
@@ -19,6 +19,17 @@ struct FixedArray
     T m_data[SIZE];
 };
 
+// Typically defined in cuda.h, but we can't ship/rely on that, so just define here
+typedef unsigned long long CUtexObject;                   
+typedef unsigned long long CUsurfObject;                  
+
+// On CUDA sampler state is actually bound up with the texture object. We have a SamplerState type, 
```

Comment:
One minor optimization here would be to make `SamplerState` be handled as a zero-byte type during layout for CUDA, and then rely on the rest of the code generation process to strip out the zero-byte field. This is handled on the HLSL/GLSL paths (although it might be part of type legalization...). Basically, empty `struct`s and fields of type `void` may appear in the IR, but by the time we go to generate code they get eliminated and don't appear in the output HLSL/GLSL.

For HLSL/GLSL this is specifically an optimization to deal with a `struct` that might have resource-type fields in it, but nothing else. If we didn't eliminate empty `struct` types then such a type would use up 16 bytes of constant-buffer space, which is just silly (also, empty `struct` types are explicitly illegal in GLSL).

This might not be as easy to do for CUDA, since code wouldn't pass through the current type legalization logic, but the idea of treating `SamplerState` as a zero-sized type is sound.

Furthermore, Slang should (eventually) be consistent in its handling of zero-size types and empty `struct`s. I'd like us to get to a point where `void` is always treated as an ordinary type that just happens to consume zero bytes. That choice could make the front-end and IR a bit more consistent, and allows `void` to be used as a "unit" type, which can be helpful for generic programming.

Reply:
We actually need to have a higher-level conversation about `SamplerState` in CUDA, though.

Both CUDA and OpenGL have the property that there is no separate `SamplerState` concept, and instead sampling state is built into texture types there. This is at odds with the behavior for D3D 11/12, where textures and samplers are always split. Of course, Vulkan supports *both* conventions, further complicating things.

There are three main ways we can try to handle the disparity:

1. We can try to model things "correctly" by having both the `Texture2D` and `SamplerState` types, along with an explicit `TextureSampler2D` type. OpenGL and CUDA targets would only support the "combined" type. D3D11/12 would support the "split" types, and would treat the combined types as if there were a two-element `struct` of a texture and a sampler. Vulkan would support each case explicitly, since they are all meaningful and different in SPIR-V.

2. As an alternative to (1) or in addition to it, we could decide that on OpenGL and CUDA, the split `Texture2D` and `SamplerState` types are supported, with `Texture2D` treated as equivalent to `TextureSampler2D` and `SamplerState` treated as an empty type. This is pretty much what you are doing right now, and it might be argued to be "principle of least surprise" for developers who are trying to write portable code. The one big gotcha here is that shader code that uses the same texture with two different samplers cannot be executed with correct semantics on OpenGL/CUDA, no matter what the application binds.

3. There's one more option that can combine with (1). This is easiest to think of in terms of OpenGL. We can treat the `Texture2D` and `SamplerState` types *as if* they are supported by OpenGL in the front end (and report them as having their own binding slots/regsisters), then during code generation we do an IR analysis path to determine which texture+sampler combinations are used (or could be used) in the generated code, and generate the "real" paired texture/sampler bindings from that. The application would then have to query some side-band data associated with each kernel to tell it how to fill in the actual OpenGL texture/sampler binding slots based on the "virtual" texture and sampler registers that the front-end reported. This option could support more flexible use of textures+samplers in the code (including one texture being used with multiple samplers), but requires a lot more integration work for an application. It also runs into serious problems in cases where the possible texture/sampler pairings can't be resolved statically, or introduce a very large space of combinations (e.g., you combine `largeTextureArray[i]` with `largeSamplerArray[j]` requiring O(N*M) combined bindings!).

In the past @csyonghe advocated for option (3) as the Right Way to do things for OpenGL, and I think that's academically correct, but maybe not practical to implement. Going with (1 + 2) is probably the simplest option for now, but I want to make sure we think about it because doing (2) rules out doing (3) at a later date without breaking compatibility.